File: Codes/Programmers/자동완성.js
class Node {
  constructor(key) {
    this.key = key;
    this.cnt = 1;
    this.children = {};
  }
}

class Trie {
  constructor() {
    this.head = new Node("");
  }

  insert(string) {
    let node = this.head;
    for (let char of string) {
      if (Object.keys(node.children).includes(char)) {
        node.children[char].cnt++;
      } else {
        node.children[char] = new Node(char);
      }
      node = node.children[char];
    }
  }

  search(string) {
    let node = this.head;
    for (let i = 0; i < string.length; i++) {
      const char = string.charAt(i);
      node = node.children[char];
      if (node.cnt == 1) {
        return i;
      }
    }
    return string.length - 1;
  }
}

function solution(words) {
  var answer = 0;
  const trie = new Trie();
  words.forEach((word) => {
    trie.insert(word);
  });

  words.forEach((word) => {
    answer += trie.search(word) + 1;
  });

  return answer;
}

File: Codes/Programmers/자동완성.py
class Node:
    def __init__(self, key):
        self.key = key
        self.cnt = 1
        self.children = {}


class Trie:
    def __init__(self):
        self.head = Node("")

    def insert(self, string):
        node = self.head
        for char in string:
            if char not in node.children:
                node.children[char] = Node(char)
            else:
                node.children[char].cnt += 1
            node = node.children[char]

    def search(self, string):
        node = self.head
        for i, char in enumerate(string):
            node = node.children[char]
            if node.cnt == 1:
                return i
        return i


def solution(words):
    answer = 0
    trie = Trie()
    for word in words:
        trie.insert(word)

    for word in words:
        answer += trie.search(word) + 1

    return answer
